module/mempool/stdmap/backdata/arraylinkedlist/entitiesLinkedList.go

```diff
@@ -0,0 +1,350 @@
+		e.free.tail.setUndefined()
+	}
+
+	// clears pointers of claimed head
+	e.values[oldFreeHeadIndex].node.next.setUndefined()
+	e.values[oldFreeHeadIndex].node.prev.setUndefined()
+
+	return oldFreeHeadIndex
+}
+
+// Rem removes entity corresponding to given sliceIndex from the list.
+func (e *EntityDoubleLinkedList) Rem(sliceIndex EIndex) {
+	e.invalidateEntityAtIndex(sliceIndex)
+}
+
+// invalidateRandomEntity invalidates the given sliceIndex in the linked list by
+// removing its corresponding linked-list node from used linked list, and appending
+// it to the tail of free list. It also removes the entity that invalidated node is presenting,
+func (e *EntityDoubleLinkedList) invalidateEntityAtIndex(sliceIndex EIndex) {
+	prev := e.values[sliceIndex].node.prev
+	next := e.values[sliceIndex].node.next
+
+	if sliceIndex != e.used.head.sliceIndex() && sliceIndex != e.used.tail.sliceIndex() {
+		// links next and prev elements for non-head and non-tail element
+		e.link(prev, next.sliceIndex())
+	}
+
+	if sliceIndex == e.used.head.sliceIndex() {
+		// invalidating used head
+		// moves head forward
+		oldUsedHead, _ := e.getHeads()
+		e.used.head = oldUsedHead.node.next
+		// new head should point to an undefined prev,
+		// but we first check if list is not empty, i.e.,
+		// head itself is not undefined.
+		if !e.used.head.isUndefined() {
+			usedHead, _ := e.getHeads()
+			usedHead.node.prev.setUndefined()
+		}
+	}
+
+	if sliceIndex == e.used.tail.sliceIndex() {
+		// invalidating used tail
+		// moves tail backward
+		oldUsedTail, _ := e.getTails()
+		e.used.tail = oldUsedTail.node.prev
+		// new head should point tail to an undefined next,
+		// but we first check if list is not empty, i.e.,
+		// tail itself is not undefined.
+		if !e.used.tail.isUndefined() {
+			usedTail, _ := e.getTails()
+			usedTail.node.next.setUndefined()
+		}
+	}
+
+	// invalidates entity and adds it to free entities.
+	e.values[sliceIndex].id = flow.ZeroID
+	e.values[sliceIndex].entity = nil
+	e.values[sliceIndex].node.next.setUndefined()
+	e.values[sliceIndex].node.prev.setUndefined()
+
+	e.appendToFreeList(sliceIndex)
+
+	// decrements Size
+	e.total--
+}
+
+// appendToFreeList appends linked-list node represented by sliceIndex to tail of free list.
+func (e *EntityDoubleLinkedList) appendToFreeList(sliceIndex EIndex) {
+	if e.free.head.isUndefined() {
+		// free list is empty
+		e.free.head.setPointer(sliceIndex)
+		e.free.tail.setPointer(sliceIndex)
+		return
+	}
+
+	// appends to the tail, and updates the tail
+	e.link(e.free.tail, sliceIndex)
+	e.free.tail.setPointer(sliceIndex)
+}
+
+// isInvalidate returns true if linked-list node represented by sliceIndex does not contain
+// a valid entity.
+func (e EntityDoubleLinkedList) isInvalidated(sliceIndex EIndex) bool {
+	if e.values[sliceIndex].id != flow.ZeroID {
+		return false
+	}
+
+	if e.values[sliceIndex].entity != nil {
+		return false
+	}
+
+	return true
+}
```

Comment:
If abstracted as suggested here [1], then this function would become something like `heroPool.GetStateByIndex(index uint32)`, i.e. we ask a pool item what its state is... and in this implementation it would either be `stateFree` or `stateUsed`...

[1] https://www.notion.so/dapperlabs/Intro-to-heap-friendly-hero-structures-cd613aba6f4244f88e58fec1f222f476

Reply:
As mentioned earlier, issued up the state-based abstraction here to address separately: dapperlabs/flow-go#6083